File: Spark学习/SparkSQL/src/scala/RDD2DataFrameReflection.scala
package cn.spark.study.sql

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.sql.SQLContext


/**
 * 如果要用scala开发spark程序
 * 然后在其中，还要实现基于反射的RDD到DataFrame的转换，就必须得用object extends App的方式
 * 不能用def main()方法的方式，来运行程序，否则就会报no typetag for ...class的错误
 * @author Administrator
 */
object RDD2DataFrameReflection extends App {
  
  val conf = new SparkConf()
      .setMaster("local")  
      .setAppName("RDD2DataFrameReflection")  
  val sc = new SparkContext(conf)
  val sqlContext = new SQLContext(sc)
  
  // 在Scala中使用反射方式，进行RDD到DataFrame的转换，需要手动导入一个隐式转换
  import sqlContext.implicits._  
  
  case class Student(id: Int, name: String, age: Int)
  
  // 这里其实就是一个普通的，元素为case class的RDD
  // 直接对它使用toDF()方法，即可转换为DataFrame
  val studentDF = sc.textFile("C://Users//htfeng//Desktop//students.txt", 1)
      .map { line => line.split(",") }
      .map { arr => Student(arr(0).trim().toInt, arr(1), arr(2).trim().toInt) }
      .toDF()    
    
  studentDF.registerTempTable("students")  
  
  val teenagerDF = sqlContext.sql("select * from students where age<=18")  
  
  val teenagerRDD = teenagerDF.rdd
  
  // 在scala中，row中的数据的顺序，反而是按照我们期望的来排列的，这个跟java是不一样的哦
  teenagerRDD.map { row => Student(row(0).toString().toInt, row(1).toString(), row(2).toString().toInt) }
      .collect()
      .foreach { stu => println(stu.id + ":" + stu.name + ":" + stu.age) }  
      
  // 在scala中，对row的使用，比java中的row的使用，更加丰富
  // 在scala中，可以用row的getAs()方法，获取指定列名的列
  teenagerRDD.map { row => Student(row.getAs[Int]("id"), row.getAs[String]("name"), row.getAs[Int]("age")) }
      .collect()
      .foreach { stu => println(stu.id + ":" + stu.name + ":" + stu.age) }
  
  // 还可以通过row的getValuesMap()方法，获取指定几列的值，返回的是个map
  val studentRDD = teenagerRDD.map { row => {
      val map = row.getValuesMap[Any](Array("id", "name", "age")); 
      Student(map("id").toString().toInt, map("name").toString(), map("age").toString().toInt)    
    } 
  }  
  studentRDD.collect().foreach { stu => println(stu.id + ":" + stu.name + ":" + stu.age) }  
}

File: Spark学习/SparkSQL/src/java/RDD2DataFrameReflection.java
package cn.spark.study.sql;

import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;

public class RDD2DataFrameReflection {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf()
				.setAppName("RDD2DataFrameReflection")
				.setMaster("local");
		
		JavaSparkContext sc = new JavaSparkContext(conf);
		SQLContext sqlContext = new SQLContext(sc);
		
		JavaRDD<String> lines = sc.textFile("C://Users//htfeng//Desktop//students.txt");
		
		JavaRDD<Student> students = lines.map(new Function<String, Student>() {
			private static final long serialVersionUID = 1L;

			@Override
			public Student call(String line) throws Exception {
				String[] lineSplited = line.split(",");  
				Student stu = new Student();
				stu.setId(Integer.valueOf(lineSplited[0].trim()));  
				stu.setName(lineSplited[1]);  
				stu.setAge(Integer.valueOf(lineSplited[2].trim())); 
				return stu;
			}
		});
		
		// 使用反射方式，将RDD转换为DataFrame
		// 将Student.class传入进去，其实就是用反射的方式来创建DataFrame
		// 因为Student.class本身就是反射的一个应用
		// 然后底层还得通过对Student Class进行反射，来获取其中的field
		// 这里要求，JavaBean必须实现Serializable接口，是可序列化的
		
		DataFrame studentDF = sqlContext.createDataFrame(students, Student.class);
		
		// 拿到了一个DataFrame之后，就可以将其注册为一个临时表，然后针对其中的数据执行SQL语句
		studentDF.registerTempTable("students");
		
		// 针对students临时表执行SQL语句，查询年龄小于等于18岁的学生，就是teenageer
		DataFrame teenagerDF = sqlContext.sql("select * from students where age<=18");
		
		// 将查询出来的DataFrame，再次转换为RDD
		JavaRDD<Row> teenagerRDD = teenagerDF.javaRDD();
		
		// 将RDD中的数据，进行映射，映射为Student
		JavaRDD<Student> teenagerStudentRDD = teenagerRDD.map(new Function<Row, Student>() {

			private static final long serialVersionUID = 1L;

			@Override
			public Student call(Row row) throws Exception {
				// TODO Auto-generated method stub
				Student stu = new Student();
				stu.setAge(row.getInt(0));
				stu.setId(row.getInt(1));
				stu.setName(row.getString(2));
				return stu;
			}
		
		});
		
		// 将数据collect回来，打印出来
		List<Student> studentList = teenagerStudentRDD.collect();
		for(Student stu: studentList) {
			System.out.println(stu.getId() + " " + stu.getAge() + " " + stu.getName());
		}
	}
	
}
